test/Passes/lower-packs-and-unpacks-without-transpose.mlir

```diff
@@ -0,0 +1,125 @@
+#map1 = affine_map<(d0, d1, d2, d3, d4, d5) -> (d1, d2, d5, d4)>
+#map2 = affine_map<(d0, d1, d2, d3, d4, d5) -> (d0, d1, d3, d4)>
+func.func @revert_all_packing(%arg0: tensor<128x512xf32>, %arg1: tensor<512x256xf32>, %arg2: tensor<128x256xf32>) -> tensor<128x256xf32> {
+  %0 = tensor.empty() : tensor<4x16x32x32xf32>
+  %pack = tensor.pack %arg0 outer_dims_perm = [0, 1] inner_dims_pos = [0, 1] inner_tiles = [32, 32] into %0 : tensor<128x512xf32> -> tensor<4x16x32x32xf32>
+  %1 = tensor.empty() : tensor<8x16x32x32xf32>
+  %pack_0 = tensor.pack %arg1 outer_dims_perm = [1, 0] inner_dims_pos = [0, 1] inner_tiles = [32, 32] into %1 : tensor<512x256xf32> -> tensor<8x16x32x32xf32>
+  %2 = tensor.empty() : tensor<4x8x32x32xf32>
+  %pack_1 = tensor.pack %arg2 inner_dims_pos = [0, 1] inner_tiles = [32, 32] into %2 : tensor<128x256xf32> -> tensor<4x8x32x32xf32>
+  %3 = linalg.generic {indexing_maps = [#map, #map1, #map2], iterator_types = ["parallel", "parallel", "reduction", "parallel", "parallel", "reduction"]} ins(%pack, %pack_0 : tensor<4x16x32x32xf32>, tensor<8x16x32x32xf32>) outs(%pack_1 : tensor<4x8x32x32xf32>) {
+  ^bb0(%in: f32, %in_2: f32, %out: f32):
+    %4 = arith.mulf %in, %in_2 : f32
+    %5 = arith.addf %out, %4 : f32
+    linalg.yield %5 : f32
+  } -> tensor<4x8x32x32xf32>
+  %unpack = tensor.unpack %3 inner_dims_pos = [0, 1] inner_tiles = [32, 32] into %arg2 : tensor<4x8x32x32xf32> -> tensor<128x256xf32>
+  return %unpack : tensor<128x256xf32>
+}
+
+// CHECK-LABEL: func.func @revert_all_packing(
+// CHECK-SAME: %[[ARG0:.+]]: tensor<128x512xf32>,
+// CHECK-SAME: %[[ARG1:.+]]: tensor<512x256xf32>,
+// CHECK-SAME: %[[ARG2:.+]]: tensor<128x256xf32>)
+// CHECK-SAME: -> tensor<128x256xf32>
+  // CHECK: %[[EXP0:.+]] = tensor.expand_shape %[[ARG0]] {{\[}}[0, 1], [2, 3]{{\]}} output_shape [4, 32, 16, 32] : tensor<128x512xf32> into tensor<4x32x16x32xf32>
+  // CHECK: %[[EXP1:.+]] = tensor.expand_shape %[[ARG1]] {{\[}}[0, 1], [2, 3]{{\]}} output_shape [16, 32, 8, 32] : tensor<512x256xf32> into tensor<16x32x8x32xf32>
+  // CHECK: %[[EXP2:.+]] = tensor.expand_shape %[[ARG2]] {{\[}}[0, 1], [2, 3]{{\]}} output_shape [4, 32, 8, 32] : tensor<128x256xf32> into tensor<4x32x8x32xf32>
+  // CHECK: %[[RES:.+]] = linalg.generic {{.*}} ins(%[[EXP0]], %[[EXP1]] : tensor<4x32x16x32xf32>, tensor<16x32x8x32xf32>) outs(%[[EXP2]] : tensor<4x32x8x32xf32>)
+  // CHECK: %[[COL:.+]] = tensor.collapse_shape %[[RES]] {{\[}}[0, 1], [2, 3]{{\]}} : tensor<4x32x8x32xf32> into tensor<128x256xf32>
+  // CHECK: return %[[COL]]
+
+// -----
+
+#map = affine_map<(d0, d1, d2, d3, d4, d5) -> (d0, d2, d3, d5)>
+#map1 = affine_map<(d0, d1, d2, d3, d4, d5) -> (d1, d2, d5, d4)>
+#map2 = affine_map<(d0, d1, d2, d3, d4, d5) -> (d0, d1, d3, d4)>
+func.func @only_keep_constant_packed(%arg0: tensor<128x512xf32>, %arg1: tensor<128x256xf32>) -> tensor<128x256xf32> {
+  %cst = arith.constant dense<1.000000e-03> : tensor<8x16x32x32xf32>
```

Comment:
I think this constant should have a pack op. Prepacked one doesn't test anything.

Reply:
The pre-transform IR for this test is just `-pack-matmul` applied to a matmul where one of the arguments is a constant. The difference between this test and the test after it is exactly the difference between the pack having already been applied or not.

If this test is not useful (as has more to do with the integration between `-pack-matmul` and the new pass, let me know.

Reply:
Converted it into actually running `-pack-matmul` on a matmul operating on a constant . This has a bit of integration test flavour, but not on the same scale as most tests in `test/Integration`.

Reply:
That makes tests more unpredictable as unrelated tests start breaking when one of the passes changes due to coupling.

IMO, it's nice to have one (perhaps cleaned up) matmul test case as that's what we care about, otherwise you can verify edge cases on much simpler examples.

Reply:
Only the first test case uses `-pack-matmul` when run - the other tests use IR obtained by manually running `-pack-matmul`. It is also as simple as we can go to test a matmul with a constant and covers the situation that we care about. Hence I believe the tests already are as you suggest.

I think in this case it is also not a bad idea to have such a "coupled" test - though possibly we need to move it to  `test/Integration` - as we would like to know when changes to either pass results in this part of the pipeline no longer doing what we expect.

Reply:
> Only the first test case uses -pack-matmul when run - the other tests use IR obtained by manually running -pack-matmul

Then the one obtained from a manual run should be enough. Otherwise, the pass can be tested with more readable 1 input 1 output cases + maybe sth for multiple results if you wanna stress that.

> I think in this case it is also not a bad idea to have such a "coupled" test - though possibly we need to move it to test/Integration

It's one of those things that are good in theory but super annoying in practice. And that's what we've started with as seen by the current collection of integration and bundle tests. In reality, they end up being brittle and whenever sth breaks I try to at least simplify or tweak them. Although, most could be removed with no loss.

I'd be all for integration tests that validate full execution by examining actual computed non-trivial results to double check correctness. I think it's sth that we lack currently.

> as we would like to know when changes to either pass results in this part of the pipeline no longer doing what we expect.

I think it'd be more helpful to verify behavior of the whole pipeline on more known kernels.
Overall, that is good goal but I don't think manually made "test pipelines" truly achieve it.

Reply:
But all in all, it's subjective. I'm not gonna block on that - if you prefer to have such multi pass example, I'm fine with it in integration tests.

Reply:
Made an integration test to test functional correctness in case of a matmul with a constant as argument. Replaced the `-pack-matmul` test case with a single argument linalg.generic test.